From 00ef0a7bb80adfdf76b33cf975fb5c4a63b5634c Mon Sep 17 00:00:00 2001
From: Alexander Hansen <alexander.hansen@9elements.com>
Date: Wed, 1 Jul 2026 15:24:18 +0200
Subject: [PATCH] allow for updating spi partitions

Some SPI flashes are partitioned to allow for A/B firmware updates
without requiring multiple chips.

Devicetree example:
```
flash@1 {
compatible = "jedec,spi-nor";
reg = <1>;
partitions {
 compatible = "fixed-partitions";
 #address-cells = <1>;
 #size-cells = <1>;

 host-prime@0 {
  label = "host-prime";
  reg = <0x0 0x2000000>;
 };
 host-second@2000000 {
  label = "host-second";
  reg = <0x2000000 0x2000000>;
 };
};
};
```

This shows up in sysfs like below:

```
root@hpe-proliant-g11:~# ls /sys/class/spi_master/spi1/spi1.1/mtd/
mtd11    mtd11ro  mtd12    mtd12ro
root@hpe-proliant-g11:~# cat /sys/class/spi_master/spi1/spi1.1/mtd/mtd11/name
host-prime
root@hpe-proliant-g11:~# cat /sys/class/spi_master/spi1/spi1.1/mtd/mtd12/name
host-second
```

Adjust the code to allow for configuring the partition to be updated.
In case no partition is configured (empty string property value), the
update should work as before.

Tested: confirmed that the correct flash partition was chosen during the
update process

```
root@hpe-proliant-g11:~# ps
...
432 root      2528 R    flashcp -v /tmp/spi-device-image-2843.bin /dev/mtd11

root@hpe-proliant-g11:~# cat /sys/class/spi_master/spi1/spi1.1/mtd/mtd11/name
host-prime
```

Upstream-Status: Submitted [Gerrit, https://gerrit.openbmc.org/c/openbmc/phosphor-bmc-code-mgmt/+/91916]
Change-Id: I42eaf85c925cde3d5029fb6b9165d2572ca13978
Signed-off-by: Alexander Hansen <alexander.hansen@9elements.com>
---
 spi-flash/bios/bios_device.cpp     |  7 +++---
 spi-flash/bios/bios_device.hpp     |  5 +++-
 spi-flash/spi_device.cpp           | 40 ++++++++++++++++++++++++------
 spi-flash/spi_device.hpp           |  5 +++-
 spi-flash/spi_factory.cpp          |  9 ++++---
 spi-flash/spi_factory.hpp          |  4 ++-
 spi-flash/spi_software_manager.cpp | 11 +++++++-
 7 files changed, 63 insertions(+), 18 deletions(-)

diff --git a/spi-flash/bios/bios_device.cpp b/spi-flash/bios/bios_device.cpp
index f46e6d33..04132a7e 100644
--- a/spi-flash/bios/bios_device.cpp
+++ b/spi-flash/bios/bios_device.cpp
@@ -30,11 +30,12 @@ enum FlashTool getBiosFlashTool(const std::string& configType)
 
 BIOSDevice::BIOSDevice(sdbusplus::async::context& ctx,
                        uint64_t spiControllerIndex, uint64_t spiDeviceIndex,
-                       bool dryRun, const std::vector<std::string>& gpioLinesIn,
+                       const std::optional<std::string>& partition, bool dryRun,
+                       const std::vector<std::string>& gpioLinesIn,
                        const std::vector<bool>& gpioValuesIn,
                        SoftwareConfig& config, SoftwareManager* parent) :
-    SPIDevice(ctx, spiControllerIndex, spiDeviceIndex, dryRun, gpioLinesIn,
-              gpioValuesIn, config, parent, flashLayoutFlat,
+    SPIDevice(ctx, spiControllerIndex, spiDeviceIndex, partition, dryRun,
+              gpioLinesIn, gpioValuesIn, config, parent, flashLayoutFlat,
               getBiosFlashTool(config.configType)),
     versionWatch(ctx, biosVersionDirPath, *this)
 {
diff --git a/spi-flash/bios/bios_device.hpp b/spi-flash/bios/bios_device.hpp
index f3c661f9..91ad8e57 100644
--- a/spi-flash/bios/bios_device.hpp
+++ b/spi-flash/bios/bios_device.hpp
@@ -3,6 +3,8 @@
 #include "common/include/host_power.hpp"
 #include "spi_device.hpp"
 
+#include <optional>
+
 using namespace phosphor::software;
 using namespace phosphor::software::manager;
 using namespace phosphor::software::host_power;
@@ -11,7 +13,8 @@ class BIOSDevice : public SPIDevice
 {
   public:
     BIOSDevice(sdbusplus::async::context& ctx, uint64_t spiControllerIndex,
-               uint64_t spiDeviceIndex, bool dryRun,
+               uint64_t spiDeviceIndex,
+               const std::optional<std::string>& partition, bool dryRun,
                const std::vector<std::string>& gpioLinesIn,
                const std::vector<bool>& gpioValuesIn, SoftwareConfig& config,
                SoftwareManager* parent);
diff --git a/spi-flash/spi_device.cpp b/spi-flash/spi_device.cpp
index f5910a09..150e0cf8 100644
--- a/spi-flash/spi_device.cpp
+++ b/spi-flash/spi_device.cpp
@@ -56,18 +56,18 @@ static std::optional<std::string> getSPIDevAddr(uint64_t spiControllerIndex)
     return std::nullopt;
 }
 
-SPIDevice::SPIDevice(sdbusplus::async::context& ctx,
-                     uint64_t spiControllerIndex, uint64_t spiDeviceIndex,
-                     bool dryRun, const std::vector<std::string>& gpioLinesIn,
-                     const std::vector<bool>& gpioValuesIn,
-                     SoftwareConfig& config, SoftwareManager* parent,
-                     enum FlashLayout layout, enum FlashTool tool) :
+SPIDevice::SPIDevice(
+    sdbusplus::async::context& ctx, uint64_t spiControllerIndex,
+    uint64_t spiDeviceIndex, const std::optional<std::string>& partition,
+    bool dryRun, const std::vector<std::string>& gpioLinesIn,
+    const std::vector<bool>& gpioValuesIn, SoftwareConfig& config,
+    SoftwareManager* parent, enum FlashLayout layout, enum FlashTool tool) :
     Device(ctx, config, parent,
            {RequestedApplyTimes::Immediate, RequestedApplyTimes::OnReset}),
     dryRun(dryRun), gpioLines(gpioLinesIn),
     gpioValues(gpioValuesIn.begin(), gpioValuesIn.end()),
     spiControllerIndex(spiControllerIndex), spiDeviceIndex(spiDeviceIndex),
-    layout(layout), tool(tool)
+    partition(partition), layout(layout), tool(tool)
 {
     auto optAddr = getSPIDevAddr(spiControllerIndex);
 
@@ -461,6 +461,20 @@ sdbusplus::async::task<bool> SPIDevice::writeSPIFlashDefault(
     co_return true;
 }
 
+// @returns empty string if no name was found
+static std::string getMTDName(const std::filesystem::path& mtdPath)
+{
+    const std::filesystem::path namePath = mtdPath / "name";
+    std::string partitionName = "";
+
+    if (std::filesystem::exists(namePath))
+    {
+        std::ifstream nameFile(namePath);
+        nameFile >> partitionName;
+    }
+    return partitionName;
+}
+
 std::optional<std::string> SPIDevice::getMTDDevicePath() const
 {
     const std::string spiPath =
@@ -478,7 +492,17 @@ std::optional<std::string> SPIDevice::getMTDDevicePath() const
     {
         const std::string mtdName = entry.path().filename().string();
 
-        if (mtdName.starts_with("mtd") && !mtdName.ends_with("ro"))
+        if (!mtdName.starts_with("mtd") || mtdName.ends_with("ro"))
+        {
+            continue;
+        }
+
+        if (!partition.has_value())
+        {
+            return "/dev/" + mtdName;
+        }
+
+        if (getMTDName(entry.path()) == partition.value())
         {
             return "/dev/" + mtdName;
         }
diff --git a/spi-flash/spi_device.hpp b/spi-flash/spi_device.hpp
index f7778663..774045e0 100644
--- a/spi-flash/spi_device.hpp
+++ b/spi-flash/spi_device.hpp
@@ -35,7 +35,8 @@ class SPIDevice : public Device
   public:
     using Device::softwareCurrent;
     SPIDevice(sdbusplus::async::context& ctx, uint64_t spiControllerIndex,
-              uint64_t spiDeviceIndex, bool dryRun,
+              uint64_t spiDeviceIndex,
+              const std::optional<std::string>& partition, bool dryRun,
               const std::vector<std::string>& gpioLinesIn,
               const std::vector<bool>& gpioValuesIn, SoftwareConfig& config,
               SoftwareManager* parent, enum FlashLayout layout,
@@ -66,6 +67,8 @@ class SPIDevice : public Device
 
     uint64_t spiControllerIndex;
     uint64_t spiDeviceIndex;
+    // optional name of the flash partition as listed in /dev/mtd/by-name/
+    std::optional<std::string> partition;
 
     // e.g. "1e631000.spi"
     std::string spiDev;
diff --git a/spi-flash/spi_factory.cpp b/spi-flash/spi_factory.cpp
index 0a0dca3c..49aa507a 100644
--- a/spi-flash/spi_factory.cpp
+++ b/spi-flash/spi_factory.cpp
@@ -2,6 +2,8 @@
 
 #include "bios/bios_device.hpp"
 
+#include <optional>
+
 namespace phosphor::software::manager
 {
 
@@ -13,7 +15,8 @@ SPIFactory& SPIFactory::instance()
 
 std::unique_ptr<SPIDevice> SPIFactory::create(
     const std::string& chipType, sdbusplus::async::context& ctx,
-    uint64_t spiControllerIndex, uint64_t spiDeviceIndex, bool dryRun,
+    uint64_t spiControllerIndex, uint64_t spiDeviceIndex,
+    const std::optional<std::string>& partition, bool dryRun,
     const std::vector<std::string>& names, const std::vector<bool>& values,
     SoftwareConfig& config, SoftwareManager* parent)
 {
@@ -23,8 +26,8 @@ std::unique_ptr<SPIDevice> SPIFactory::create(
         try
         {
             return std::make_unique<BIOSDevice>(
-                ctx, spiControllerIndex, spiDeviceIndex, dryRun, names, values,
-                config, parent);
+                ctx, spiControllerIndex, spiDeviceIndex, partition, dryRun,
+                names, values, config, parent);
         }
         catch (const std::exception& e)
         {
diff --git a/spi-flash/spi_factory.hpp b/spi-flash/spi_factory.hpp
index 8ad4418a..0e51c837 100644
--- a/spi-flash/spi_factory.hpp
+++ b/spi-flash/spi_factory.hpp
@@ -6,6 +6,7 @@
 #include <sdbusplus/async.hpp>
 
 #include <memory>
+#include <optional>
 #include <string>
 #include <unordered_map>
 #include <vector>
@@ -50,7 +51,8 @@ class SPIFactory
 
     static std::unique_ptr<SPIDevice> create(
         const std::string& chipType, sdbusplus::async::context& ctx,
-        uint64_t spiControllerIndex, uint64_t spiDeviceIndex, bool dryRun,
+        uint64_t spiControllerIndex, uint64_t spiDeviceIndex,
+        const std::optional<std::string>& partition, bool dryRun,
         const std::vector<std::string>& names, const std::vector<bool>& values,
         SoftwareConfig& config, SoftwareManager* parent);
 
diff --git a/spi-flash/spi_software_manager.cpp b/spi-flash/spi_software_manager.cpp
index 5e655a74..be059d82 100644
--- a/spi-flash/spi_software_manager.cpp
+++ b/spi-flash/spi_software_manager.cpp
@@ -45,6 +45,15 @@ sdbusplus::async::task<bool> SPISoftwareManager::initDevice(
         co_return false;
     }
 
+    std::optional<std::string> partition =
+        co_await dbusGetRequiredProperty<std::string>(ctx, service, path,
+                                                      configIface, "Partition");
+
+    if (!partition.has_value())
+    {
+        debug("no partitions configured");
+    }
+
     const std::string configIfaceMux = configIface + ".MuxOutputs";
 
     std::vector<std::string> names;
@@ -76,7 +85,7 @@ sdbusplus::async::task<bool> SPISoftwareManager::initDevice(
 
     auto spiDevice = SPIFactory::instance().create(
         chipType, ctx, spiControllerIndex.value(), spiDeviceIndex.value(),
-        dryRun, names, values, config, this);
+        partition, dryRun, names, values, config, this);
 
     if (spiDevice == nullptr)
     {
-- 
2.55.0

